src/main/java/org/thingsboard/gateway/service/MqttMessageServiceImpl.java

```diff
@@ -0,0 +1,104 @@
+package org.thingsboard.gateway.service;
+
+import lombok.extern.slf4j.Slf4j;
+import org.eclipse.paho.client.mqttv3.MqttMessage;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.data.domain.PageRequest;
+import org.springframework.data.domain.Sort;
+import org.springframework.stereotype.Service;
+import org.springframework.transaction.annotation.Transactional;
+import org.thingsboard.gateway.dao.PersistentMqttMessage;
+import org.thingsboard.gateway.dao.PersistentMqttMessageRepository;
+import org.thingsboard.gateway.service.conf.TbPersistenceConfiguration;
+
+import javax.annotation.PostConstruct;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.UUID;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.function.Consumer;
+
+/**
+ * Created by Valerii Sosliuk on 12/28/2017.
+ */
+@Service
+@Slf4j
+public class MqttMessageServiceImpl implements MqttMessageService {
+
+    @Autowired
+    private PersistentMqttMessageRepository messageRepository;
+
+    @Autowired
+    private TbPersistenceConfiguration persistence;
+
+    private Map<UUID, Consumer<PersistentMqttMessage>> successCallbacks;
```

Comment:
ConcurrentMap interface may be useful.

Reply:
Done, although I don't use anything else than get() and put()